### DISCUSSION TITLE
[linux] Visual bugs with Nouveau on Nvidia 2070 (Turing generation)

### DISCUSSION THREAD

**Discussion by TechniKris** [ppy/osu#30290] [Q&A] [answered]
### Type

Cosmetic

### Bug description

There are multiple visual bugs, all seemingly related to circular, pie chart-shaped objects.
It seems like [Mesa](https://www.mesa3d.org/)/[Nouveau](https://nouveau.freedesktop.org/) driver might be at fault.
I thought about reporting it upstream at https://gitlab.freedesktop.org/mesa/mesa but I wanted to consult here first if you'd like to diagnose it more in-depth and handle it on your own.

### Screenshots or videos

Volume dials:

![volume_dials](https://github.com/user-attachments/assets/4719761e-115b-4730-8643-bad177596909)

`LegacySongProgress` skin element:

https://github.com/user-attachments/assets/e1599e12-5622-4df4-bc29-9441699c7b66

The "hold for menu" button:

https://github.com/user-attachments/assets/213d6bc3-f9ca-417e-bc68-97d07a0a9c4e

Grade pie chart on the results screen:

![s_grade_animation](https://github.com/user-attachments/assets/ec22b0d0-becc-4591-9823-e0dcf53d4a45)
![c_grade](https://github.com/user-attachments/assets/f31d2778-f232-4c1e-aa3b-9c5e454b4188)
![d_grade](https://github.com/user-attachments/assets/e1733039-930e-47d7-b2d3-3b2828e9e61b)

2 cases (that I've found) of specific sliders on beatmaps (note the line going through the middle of the slider):

-  beatmap ID: [1264503](https://osu.ppy.sh/b/1264503) `osu://edit/00:41:982 (2)`

https://github.com/user-attachments/assets/db2ee30a-8cc8-4a6b-87bf-8bf2e4de954a

- beatmap ID: [64896](https://osu.ppy.sh/b/64896) osu://edit/02:08:082 (1) (argon and argon pro are not affected here)

https://github.com/user-attachments/assets/cbb06fb0-4832-46a9-bdb6-c7f61bb52065

- I'm pretty sure I stumbled on such a broken slider at least once more, but I don't remember the beatmap.

Beatmap preview progress bar:

https://github.com/user-attachments/assets/4e6959c8-3ef2-488e-aec4-5467eee921eb

### Affected drivers/hardware

I've performed all of the below testing with the default `GL / OpenGL` renderer, but as a sanity check I tested `Deferred / OpenGL` and... it seems to be affected as well.

The issue occurs with the [Nouveau](https://nouveau.freedesktop.org/) driver - the Linux open source OpenGL driver for Nvidia GPUs, which is used by the system by default, if there's no [Nvidia official driver](https://www.nvidia.com/drivers/) installed.
Only 1 out of 3 tested Nvidia GPUs seems to be affected:
- RTX 2070 ❗ affected
- GTX 650 ➖ not affected
- 9300M GS (tested in a different PC - laptop) ➖ not affected

Additionally:
- Nvidia's official drivers ([470.256.02](https://www.nvidia.com/en-us/drivers/details/226760/) (for 650 and 2070), [560.35.03](https://www.nvidia.com/en-us/drivers/details/230918/) (for 2070), [560.35.03 open variant](https://github.com/NVIDIA/open-gpu-kernel-modules/releases/tag/560.35.03) (for 2070)): ➖ not affected
- AMD GPU - Ryzen 5700G's integrated graphics ➖ not affected
- [Zink](https://docs.mesa3d.org/drivers/zink.html) -  using equivalent Vulkan drivers (the game was still using `GL / OpenGL` renderer), on all of the above GPUs (except 9300M GS and GTX 650, which have no- or inadequate Vulkan) ➖ not affected

Since only the RTX 2070 (Turing generation) is affected, I suspect the bug could be specifically in some kind of codepath for [GSP](https://us.download.nvidia.com/XFree86/Linux-x86_64/560.35.03/README/gsp.html)-enabled cards, which is Turing and newer, and based on that I'd presume newer cards might also be affected. I don't have any more of such cards though, so I can't verify this.

### Workarounds

[Zink](https://docs.mesa3d.org/drivers/zink.html) seems like an adequate solution here, providing good performance on top of [NVK](https://docs.mesa3d.org/drivers/nvk.html) (or any other Vulkan driver). Zink can be enabled by setting these environment variables: `__GLX_VENDOR_LIBRARY_NAME=mesa MESA_LOADER_DRIVER_OVERRIDE=zink GALLIUM_DRIVER=zink`.

### Version

I noticed these issues over the course of last couple months, at least as early as 2024.521.2
Tested in-depth and recorded examples on 2024.1009.1

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/17364445/compressed-logs.zip)
I can provide [apitrace](https://apitrace.github.io/) if needed.

**Comment by hwsmm** [comment]
You already isolated the problem and concluded that RTX2060 with nouveau GSP is the only affected case here..? There may be some applicable workarounds, but you should consult with Mesa as you mentioned at first since nouveau GSP support is still very early.

**Reply by TechniKris** [reply]
I have tried explicitly setting `nouveau.config=NvGspRm=0` kernel parameter (available to 16/20 and 30 series) to disable anything related to the GSP, but it didn't seem to make any difference in behavior (only in performance, which I think would align with the [feature matrix](https://nouveau.freedesktop.org/FeatureMatrix.html)).
I guess this discussion can at least serve as a reference for anyone else stumbling upon the issue. Listing affected cards would help confirm my assumption about the range of GPUs affected.

**Reply by Soggy-Pancake** [reply]
Getting the same rendering issues with the opengl renderer under nouveau with my 3080.

**Comment by bdach** [comment]
previously: https://github.com/ppy/osu/discussions/29645

99.9999999% nouveau/mesa-related, not really interested personally in debugging further

**Reply by TechniKris** [reply]
Yup, most likely. And in Mesa 25.1 Zink is supposed to become the default for affected GPUs ([Collabora blog post about it](https://www.collabora.com/news-and-blog/news-and-events/goodbye-nouveau-gl-hello-zink.html)), so it'll work around the issue.
I intend to keep this discussion open until Mesa 25.1 is out and I confirm the issue is gone in the default configuration.

**Reply by TechniKris** [reply]
Also:

> previously: #29645

That seems to be an unrelated issue, as it doesn't affect me.

**Comment by TechniKris** [comment, marked as answer]
As I've mentioned earlier https://github.com/ppy/osu/discussions/30290#discussioncomment-12726596 Zink has become the default OpenGL driver on Turing+ since [Mesa 25.1](https://docs.mesa3d.org/relnotes/25.1.0.html), which basically works around the entire issue for affected hardware.

> - loader/nouveau: load zink as the GL driver for turing and above.

I have confirmed it works correctly on a Plasma Wayland desktop.

Here's the relevant OpenGL system information: `$ glxinfo | grep OpenGL`
```plain
OpenGL vendor string: Mesa
OpenGL renderer string: zink Vulkan 1.4(NVIDIA GeForce RTX 2070 (NVK TU106) (MESA_NVK))
OpenGL core profile version string: 4.6 (Core Profile) Mesa 25.1.0-arch1.1
OpenGL core profile shading language version string: 4.60
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 4.6 (Compatibility Profile) Mesa 25.1.0-arch1.1
OpenGL shading language version string: 4.60
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 25.1.0-arch1.1
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
OpenGL ES profile extensions:
```
Marking this as an answer; but feel free to speak up if you see the issue still occuring on Mesa 25.1 or above.
